rust-app/src/ui/common.rs

```diff
@@ -0,0 +1,134 @@
+use crate::parser::common::{CoinType, IOTA_COIN_DECIMALS, IOTA_COIN_TYPE};
+use crate::utils::*;
+
+extern crate alloc;
+use alloc::format;
+
+use arrayvec::ArrayString;
+use arrayvec::ArrayVec;
+use either::*;
+use ledger_crypto_helpers::common::HexSlice;
+use ledger_log::info;
+
+#[cfg(target_os = "nanox")]
+pub const MESSAGE_MAX_LENGTH: usize = 1024 * 2;
+
+#[cfg(not(target_os = "nanox"))]
+pub const MESSAGE_MAX_LENGTH: usize = 1024 * 4;
+
+pub fn is_printable_ascii(data: &ArrayVec<u8, MESSAGE_MAX_LENGTH>) -> Option<&str> {
+    if data.iter().all(|&b| b > 0) {
+        core::str::from_utf8(data.as_slice()).ok()
+    } else {
+        info!("is_printable_ascii: input contains NUL char");
+        None
+    }
+}
+
+#[inline(never)]
+pub fn get_coin_and_amount_fields(
+    total_amount: u64,
+    coin_type: CoinType,
+) -> (
+    (ArrayString<32>, ArrayString<32>),
+    Either<ArrayString<8>, (ArrayString<4>, ArrayString<256>)>,
+) {
+    if let Some((ticker, decimals)) = get_known_coin_ticker(&coin_type) {
+        let (quotient, remainder_str) = get_amount_in_decimals(total_amount, decimals);
+        let v1 = format!(
+            "{} {}.{}",
+            ticker.as_str(),
+            quotient,
+            remainder_str.as_str()
+        );
+        let amount = (
+            ArrayString::from("Amount").unwrap(),
+            ArrayString::from(&v1).unwrap(),
+        );
+        (amount, Left(ticker))
+    } else {
+        let v1 = format!("{}", total_amount);
+        let amount = (
+            ArrayString::from("Raw Amount").unwrap(),
+            ArrayString::from(&v1).unwrap(),
+        );
+        let (coin_id, module, name) = coin_type;
+
+        let v2 = format!(
+            "{}::{}::{}",
+            HexSlice(&coin_id),
+            core::str::from_utf8(module.as_slice()).unwrap_or("invalid utf-8"),
+            core::str::from_utf8(name.as_slice()).unwrap_or("invalid utf-8")
+        );
+        let coin = Right((
+            ArrayString::from("Coin").unwrap(),
+            ArrayString::from(&v2).unwrap(),
+        ));
+        (amount, coin)
+    }
+}
+
+#[inline(never)]
+fn get_known_coin_ticker(coin_type: &CoinType) -> Option<(ArrayString<8>, u8)> {
+    if *coin_type == IOTA_COIN_TYPE {
+        return Some((ArrayString::from("IOTA").unwrap(), IOTA_COIN_DECIMALS));
+    }
+
+    for k in KNOWN_COINS {
+        let mut module = ArrayVec::new();
+        let _ = module.try_extend_from_slice(k.module.as_bytes());
+
+        let mut function = ArrayVec::new();
+        let _ = function.try_extend_from_slice(k.function.as_bytes());
+
+        if *coin_type == (k.coin_id, module, function) {
+            return Some((ArrayString::from(k.ticker).unwrap(), k.decimals));
+        }
+    }
+    None
+}
+
+struct KnownCoin<'a> {
+    coin_id: [u8; 32],
+    module: &'a str,
+    function: &'a str,
```

Comment:
Shouldn't this be called different, like identifier for example?

Reply:
we can call it `witness` maybe?

Reply:
Works for me